File: MIDTERM-LAB-1/Main.java
import java.io.*;
import java.util.*;

public class Main {

    public static void main(String[] args) {

        @SuppressWarnings("resource")
        Scanner input = new Scanner(System.in);
        File file;

        // ===============================
        // FILE VALIDATION LOOP
        // ===============================
        while (true) {
            System.out.print("Enter dataset file path: ");
            String path = input.nextLine().replace("\"", "");
            file = new File(path);

            if (!file.exists()) {
                System.out.println("Error: File does not exist.\n");
            } else if (!file.canRead()) {
                System.out.println("Error: File is not readable.\n");
            } else if (!path.toLowerCase().endsWith(".csv")) {
                System.out.println("Error: File is not a CSV.\n");
            } else {
                break;
            }
        }

        List<DataRecord> records = new ArrayList<>();

        // ===============================
        // FILE PROCESSING
        // ===============================
        try (BufferedReader br = new BufferedReader(new FileReader(file))) {

            @SuppressWarnings("unused")
            String header = br.readLine(); // Skip header
            String line;

            while ((line = br.readLine()) != null) {

                // Safe CSV split (handles basic commas)
                String[] parts = line.split(",(?=(?:[^\"]*\"[^\"]*\")*[^\"]*$)");

                if (parts.length < 13) continue;

                String totalSalesStr = parts[7].trim();
                String releaseDate = parts[12].trim();

                // Skip invalid rows
                if (totalSalesStr.isEmpty() || releaseDate.isEmpty())
                    continue;

                // Ensure full YYYY-MM-DD format exists
                if (releaseDate.length() < 7)
                    continue;

                try {
                    double sales = Double.parseDouble(totalSalesStr);
                    String month = releaseDate.substring(0, 7);

                    records.add(new DataRecord(month, sales));

                } catch (NumberFormatException e) {
                    // Skip rows with invalid numeric values
                }
            }

            // ===============================
            // ANALYTICS
            // ===============================
            Map<String, Double> monthlyTotals =
                    MonthlyAnalyzer.analyze(records);

            if (monthlyTotals.isEmpty()) {
                System.out.println("No valid data found.");
                return;
            }

            System.out.println("\n===== MONTHLY SALES SUMMARY =====");

            for (Map.Entry<String, Double> entry : monthlyTotals.entrySet()) {
                System.out.printf("%-10s : %.2f million\n",
                        entry.getKey(),
                        entry.getValue());
            }

            String bestMonth =
                    MonthlyAnalyzer.getBestMonth(monthlyTotals);

            System.out.println("\nBest Performing Month: " + bestMonth);

        } catch (IOException e) {
            System.out.println("Error processing file: " + e.getMessage());
        }
    }
}

File: MIDTERM-LAB-1/main.js
const fs = require('fs');
const readline = require('readline');
const { analyze, getBestMonth } = require('./analyzer');

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

function askFilePath() {
    rl.question("Enter dataset file path: ", function(path) {

        if (!fs.existsSync(path)) {
            console.log("File does not exist.");
            return askFilePath();
        }

        if (!path.toLowerCase().endsWith(".csv")) {
            console.log("File is not a CSV.");
            return askFilePath();
        }

        try {
            const data = fs.readFileSync(path, 'utf8');
            const lines = data.split('\n').slice(1);

            const records = [];

            lines.forEach(line => {
                const parts = line.split(',');

                if (parts.length < 13) return;

                const releaseDate = parts[12];
                const totalSales = parts[7];

                if (!releaseDate || !totalSales) return;

                const month = releaseDate.substring(0, 7);
                records.push({
                    month: month,
                    sales: parseFloat(totalSales)
                });
            });

            const monthlyTotals = analyze(records);

            console.log("\n===== MONTHLY SALES SUMMARY =====");
            Object.keys(monthlyTotals)
                .sort()
                .forEach(month => {
                    console.log(`${month} : ${monthlyTotals[month].toFixed(2)} million`);
                });

            const bestMonth = getBestMonth(monthlyTotals);
            console.log(`\nBest Performing Month: ${bestMonth}`);

            rl.close();

        } catch (error) {
            console.log("Error reading file:", error.message);
            askFilePath();
        }
    });
}

askFilePath();